📖 Add note about clusterctl upgrade test coverage (#8073)
kubernetes-sigs/cluster-api: docs/book/src/clusterctl/commands/upgrade.md

<aside class="note warning">

<h1>Skip-level upgrades</h1>

Cluster API only tests clusterctl upgrades from the latest minor version of each contract for a specific version 
(e.g. v0.3 (v1alpha3) => v1.3, v0.4 (v1alpha4) => v1.3 and v1.2 (v1beta1) => v1.3). While we are not testing skip-level 

> upgrades from earlier versions (e.g. v1.1 => v1.3), it should work in general.

Reviewer: > it should work in general

I wonder if this sounds a bit too vague, but I don't have better suggestions. Maybe others can weigh in?

Author: @CecileRobertMichon Kilian suggested an improvement. Does that sound better to you?

Reviewer 2: +1 to the second version
Wondering if we can simplify the first sentence by just listing what upgrade path we are currently testing; also, if I'm not wrong we are missing 1.2 --> 1.3

Author: We can add an explicit list, but that also means we have to keep the list up-to-date.

I would prefer writing it more generic than that to avoid this

> Cluster API only tests clusterctl upgrades from the latest minor version of each contract for a specific version 
> (e.g. v0.3 (v1alpha3) => v1.3, v0.4 (v1alpha4) => v1.3 and v1.2 (v1beta1) => v1.3). While we are not testing skip-level 
> upgrades from earlier versions (e.g. v1.1 => v1.3), it should work in general.

Reviewer 3: Maybe?
```suggestion
Cluster API only tests clusterctl upgrades from the latest minor version of each contract for a specific version 
(e.g. v0.3 (v1alpha3) => v1.3, v0.4 (v1alpha4) => v1.3 and v1.2 (v1beta1) => v1.3). Skip-level upgrades from earlier versions (e.g. v1.1 => v1.3) are not blocked, but are not tested. Users intending to carry out skip-level upgrades should do their own validation to ensure the operation works correctly.
```

Author: Thx. I like it!

<aside class="note warning">

> <h1>Skip-level upgrades</h1>

Reviewer 4: Should we call this "Skipping minor versions in upgrades"?

Reviewer 5: Could we also have an upgrade or link in the [tasks / upgrading clusters](https://github.com/kubernetes-sigs/cluster-api/blob/329753cae2c23d0f29a764870dc99972a50036b0/docs/book/src/tasks/upgrading-clusters.md?plain=1#L13-L14) to this?

Author: @oscr I think this is a different topic. This PR is about skipping Cluster API minor versions while what you referenced is about skipping Kubernetes minor versions.

<aside class="note warning">

<h1>Clusterctl upgrade test coverage</h1>

Cluster API only tests a subset of possible clusterctl upgrade paths as otherwise the test matrix would be overwhelming.

> Untested upgrade paths are not blocked by clusterctl and should work in general, they are just not tested. Users 

Reviewer 3: ```suggestion
Untested upgrade paths are not blocked by clusterctl but they are just not tested. Users 
```

Nit, but I'm not sure about the grey area of "it should work, but it's not tested". Better IMO just to describe the state.

Author: Let's see what others thing. I think the assurance that this is something that we expect to work out-of-the-box might be important to users.

Reviewer 2: We can keep it as it is, given that the most important part is the next part about users doing their own validation if they want to rely on untested upgrade paths
